components/engine_traits/src/table_properties.rs

```diff
@@ -25,4 +39,47 @@ pub trait TablePropertiesExt: CFHandleExt {
     }
 }
 
-pub trait TablePropertiesCollection {}
+pub trait TablePropertiesCollection<I, PKey, P, UCP>
+where
+    I: TablePropertiesCollectionIter<PKey, P, UCP>,
+    PKey: TablePropertiesKey,
+    P: TableProperties<UCP>,
+    UCP: UserCollectedProperties,
+{
+    fn iter(&self) -> I;
+
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+pub trait TablePropertiesCollectionIter<PKey, P, UCP>: Iterator<Item = (PKey, P)>
+where
+    PKey: TablePropertiesKey,
+    P: TableProperties<UCP>,
+    UCP: UserCollectedProperties,
+{
+}
+
+pub trait TablePropertiesKey: Deref<Target = str> {}
+
+pub trait TableProperties<UCP>
+where
+    UCP: UserCollectedProperties,
+{
+    fn num_entries(&self) -> u64;
+
+    fn user_collected_properties(&self) -> UCP;
+}
+
+pub trait UserCollectedProperties {
+    fn get<Q: AsRef<[u8]>>(&self, index: Q) -> Option<&[u8]>;
```

Comment:
```suggestion
    fn get(&self, index: &[u8]) -> Option<&[u8]>;
```

What about remove `Q`? I think it's simpler.

Reply:
I agree. I'll remove it.